src/optimize/watch/watch_cache.ts

```diff
@@ -0,0 +1,173 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { createHash } from 'crypto';
+import { readFile, writeFile } from 'fs';
+import { resolve } from 'path';
+import { promisify } from 'util';
+
+import del from 'del';
+import deleteEmpty from 'delete-empty';
+import globby from 'globby';
+import mkdirp from 'mkdirp';
+
+const mkdirpAsync = promisify(mkdirp);
+const readAsync = promisify(readFile);
+const writeAsync = promisify(writeFile);
+
+interface Params {
+  log: (tags: string[], data: string) => void;
+  outputPath: string;
+  dllsPath: string;
+  cachePath: string;
+}
+
+interface WatchCacheStateContent {
+  optimizerConfigSha?: string;
+  yarnLockSha?: string;
+}
+
+export class WatchCache {
+  private readonly log: Params['log'];
+  private readonly outputPath: Params['outputPath'];
+  private readonly dllsPath: Params['dllsPath'];
+  private readonly cachePath: Params['cachePath'];
+  private readonly cacheState: WatchCacheStateContent;
+  private statePath: string;
+  private diskCacheState: WatchCacheStateContent;
+  private isInitialized: boolean;
+
+  constructor(params: Params) {
+    this.log = params.log;
+    this.outputPath = params.outputPath;
+    this.dllsPath = params.dllsPath;
+    this.cachePath = params.cachePath;
+
+    this.isInitialized = false;
+    this.statePath = '';
+    this.cacheState = {};
+    this.diskCacheState = {};
+    this.cacheState.yarnLockSha = '';
+    this.cacheState.optimizerConfigSha = '';
+  }
+
+  public async tryInit() {
+    if (!this.isInitialized) {
+      this.statePath = resolve(this.outputPath, 'watch_optimizer_cache_state.json');
+      this.diskCacheState = await this.read();
+      this.cacheState.yarnLockSha = await this.buildYarnLockSha();
+      this.cacheState.optimizerConfigSha = await this.buildOptimizerConfigSha();
+      this.isInitialized = true;
+    }
+  }
+
+  public async tryReset() {
+    await this.tryInit();
+
+    if (!this.isResetNeeded()) {
+      return;
+    }
+
+    await this.reset();
+  }
+
+  public async reset() {
+    this.log(['info', 'optimize:watch_cache'], 'The optimizer watch cache will reset');
+
+    // start by deleting the state file to lower the
+    // amount of time that another process might be able to
+    // successfully read it once we decide to delete it
+    await del(this.statePath);
+
+    // first delete some empty folder that left
+    // from any previous cache reset action
+    await deleteEmpty(`${this.cachePath}`);
```

Comment:
this catch some empty folders left by globs with expanded directories that we are deleting with `await del(await globby`. I choose to run this before the instruction that causes those empty folders (so it will delete the empty folders left by the last cache delete before the current one) but I think I can try to run it for the current cache reset. I think I just choose to run it like this to prevent some runtime errors.